File: frontend/src/Services/Poll.js
import axios_instance from "@/axios/axios";

export async function FetchPolls(){
    try{
        const response = await axios_instance.get('/poll/all')
        return response.data.data.reverse();
    }catch(error){
        console.log(error)
        throw new Error(error)
    }
}

export async function FetchVoteInfo(pollId, userId){
    try{
        if(!userId){
            userId = localStorage.getItem('userId');
        }
        const params = {"user_id" : userId, "poll_id" : pollId};
        const response = await axios_instance.get('/poll/voteinfo', {params});
        return response.data.data;
    }catch(error){
        console.log(error);
        throw new Error(error);
    }
}

export async function CreatePoll(data){
    try{
        const body = data;
        const response = await axios_instance.post('/poll/create', body)
        return response.data.message;
    }catch(error){
        console.log(error);
        throw new Error(error);
    }
}

export async function FetchMyPolls(){
    try{
        const userId = localStorage.getItem('userId');
        const params = {"userId" : userId};
        const response = await axios_instance.get('/poll/mypolls', {params});
        return response.data.data.reverse();
    }catch(error){
        console.log(error);
        throw new Error(error);
    }
}

export async function FetchOnePoll(pollname){
    try{
        const params = {"poll_name" : pollname};
        const response = await axios_instance.get('/poll/getone', {params});
        return response.data.data;
    }catch(error){
        console.log(error);
        throw new Error(error);
    }
}

File: backend/Controllers/poll.py
from flask import Blueprint, request, jsonify
from Service.poll import CreatePollService, UpdatePollService, FetchAllPollService, FetchOnePollService, FetchVoteInfoService, FetchMyPollsService
import json
from urllib.parse import unquote
poll = Blueprint('poll', __name__)
@poll.route('/create', methods=['POST'])
def CreatePollController():
    try:
        formdata = request.json
        print(formdata)
        name = formdata.get('name')
        description = formdata.get('description')
        options = formdata.get('options')
        colour = formdata.get('colour')
        author_name = formdata.get('author_name')
        emailId = formdata.get('emailId')
        userId = formdata.get('userId')
        public = int(formdata.get('public'))
        data = {"public" : public,"userId" : userId, "name" : name, "description" : description, "author_name" : author_name, "emailId" : emailId, "colour" : colour, "author_name" : author_name, "options" : options}
        response = CreatePollService(data)
        return jsonify({"message" : response}), 201
    except Exception as e:
        print(e)
        return jsonify({"error": str(e)}), 400
    

@poll.route('/update', methods=['POST'])
def UpdatePollController():
    try:
        updates = request.form.get("raw")
        id = request.form.get('_id')
        update_data = json.loads(updates)
        print(update_data)
        print(id)
        response = UpdatePollService(update_data, id)
        return jsonify({"message" : response}), 200
    except Exception as e:
        print(f"An error occured in controller{e}")
        return jsonify({"error": str(e)}), 400
    

@poll.route("/all", methods=["GET"])
def FetchAllPollController():
    try:
        response = FetchAllPollService()
        return jsonify({
            "success" : True,
            "data" : response
        }), 200
    except Exception as e:
        print(e)
        return jsonify({
            "success" : False,
            "message" : "something went wrong"
        }), 400
    

@poll.route("/getone", methods=["GET"])
def FetchOnePollController():
    try:
        pollname = request.args.get("poll_name")
        pollname = unquote(pollname)
        print(pollname)
        response = FetchOnePollService(pollname)
        return jsonify({
            "success" : True,
            "data" : response
        }), 200
    except Exception as e:
        print(e, "error in controller")
        return jsonify({
            "success" : False,
            "message" : "something went wrong"
        }), 400

@poll.route("/voteinfo", methods=["GET"])
def FetchVoteInfoController():
    try:
        pollId = request.args.get("poll_id")
        userId = request.args.get("user_id")
        print(pollId, userId)
        response = FetchVoteInfoService(pollId, userId)
        return jsonify({
            "success" : True,
            "data" : response
        }), 200
    except Exception as e:
        print(e)
        return jsonify({
            "success" : False,
            "message" : e
        }), 400
        
@poll.route("/mypolls", methods=["GET"])
def FetchMyPollsController():
    try:
        userId = request.args.get("userId")
        response = FetchMyPollsService(userId)
        return jsonify({
            "success" : True,
            "data" : response
        }), 200
    except Exception as e:
        print(e)
        return jsonify({
            "success" : False,
            "message" : e
        }), 400
